spec/services/badge_granter_spec.rb

```diff
@@ -481,6 +481,16 @@
       BadgeGranter.backfill(Badge.find(Badge::GreatPost))
       expect(UserBadge.find_by(user_id: user.id, badge_id: Badge::GreatPost)).to eq(nil)
     end
+
+    it 'calls user_badge_granted on backfilled badge' do
+      post = create_post(user: user)
+      action = PostActionCreator.like(liker, post).post_action
+
+      Badge.any_instance.expects(:trigger_badge_granted_event).with(user.id).once
+      Badge.any_instance.expects(:trigger_badge_granted_event).with(liker.id).once
```

Comment:
I’m curious about the choice to mock here instead of tracking the event.

Reply:
also and maybe more importantly it would mean we need to change two tests each time we change the badge event method behavior

Reply:
Mocking here means we’re exposing the internal implementation detail though which isn’t really the concern of the test.

Reply:
So the concern here if I follow you is that we want to ensure this part of code will trigger an event, so we want to test what this part of the code will do right?

What if `:trigger_badge_granted_event` implementation changes ?

```ruby
def trigger_badge_granted_event(user_id)
  DiscourseEvent.trigger(...)
  do_something_else()
end
```

Don't we care about being sure the consequences of `do_something_else` will happen just as we care the event will be triggered? I think we do. Problem doing a mock or testing directly the event won't fail if we change the implementation.

The difference is that testing only the event in badge granter will become stalled as only testing partially what's happening, why the mock is still up to date and testing we actually call `Bade#trigger_badge_granted_event`. If you change `Badge#trigger_badge_granted_event`, you should think of changing the tests for it in Badge spec, but why would you think of this test in badge granter? Most likely you won't.

Reply:
> Problem doing a mock or testing directly the event won't fail if we change the implementation.

Yup which is the problem with the mock here. We're just testing that the method is called and not that an event has actually been triggered which is what this PR cares about. For instance, if someone decides to change the name of the event that is triggered, the spec in `BadgeGranter` will still.

>  If you change Badge#trigger_badge_granted_event, you should think of changing the tests for it in Badge spec, but why would you think of this test in badge granter? Most likely you won't.

This reinforces why we need to test for the outcome and not the internal implementation detail. As far as the test in `BadgeGranter` is concerned, it is expecting a `DiscourseEvent` with the name `:user_badge_granted` and the right arguments.

Reply:
No I really dont agree, and you dont answer what I said. You have no reason to think about changing this if the lethod changes and as a result this is a very brittle test.